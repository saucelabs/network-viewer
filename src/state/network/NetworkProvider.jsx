import React, { useEffect, useReducer, useMemo } from 'react';
import PropTypes from 'prop-types';

import { reducer, initialState as defaultState } from './reducer';
import { updateData, fetchFile, updateScrollToIndex } from './actions';
import { NetworkContext } from './Context';
import { findRequestIndex } from '../../utils';
import { ROW_ID_PREFIX } from '../../constants';

const NetworkProvider = (props) => {
  const {
    data,
    file,
    fetchOptions,
    initialState,
    scrollTimeStamp,
    scrollRequestPosition,
    autoHighlightChange,
    onDataLoaded,
    onDataError,
    onPause,
    onResume,
    onReset,
  } = props;

  const [state, dispatch] = useReducer(reducer, initialState);
  const callbacks = {
    onPause,
    onResume,
    onReset,
  };
  const value = useMemo(() => [state, dispatch, callbacks], [state]);
  const selectedReqIndex = value[0].get('selectedReqIndex');
  const requestData = value[0].get('data');
  const showReqDetail = value[0].get('showReqDetail');
  const actualData = value[0].get('actualData');
  const error = value[0].get('error');

  // Update data onChange of network data
  useEffect(() => {
    if (data && data.log && data.log.entries) {
      updateData(dispatch)(data);
    }
  }, [data]);

  // Fetch HAR file onChange of file prop
  useEffect(() => {
    if (file) {
      fetchFile(dispatch)(file, fetchOptions);
    }
  }, [file]);

  useEffect(() => {
    if (actualData.size && onDataLoaded) {
      onDataLoaded(actualData);
    }
  }, [actualData]);

  useEffect(() => {
    if (error && onDataError) {
      onDataError(error);
    }
  }, [error]);

  // Find nearby request-rowId and update scrollIndex on scrollTimeStamp receive
  useEffect(() => {
    const shouldChangeHighlight = showReqDetail ? autoHighlightChange : true;
    if (scrollTimeStamp && shouldChangeHighlight) {
      const reqIndex = findRequestIndex({
        data: requestData,
        timestamp: scrollTimeStamp,
        position: scrollRequestPosition,
      });
      if (reqIndex || reqIndex === 0) {
        updateScrollToIndex(dispatch)(requestData.get(reqIndex));
      }
    }
  }, [scrollTimeStamp, actualData]);

  // Scroll to request row onChange of scrollToIndex
  // setTimeout is required when reqDetail is visible,
  // it allows DOM to adjust before we scroll it to highlighted request
  useEffect(() => {
    if (selectedReqIndex) {
      setTimeout(() => {
        const element = document.getElementById(ROW_ID_PREFIX + selectedReqIndex);
        if (element) {
          element.scrollIntoView({
            alignToTop: true,
            behavior: 'smooth',
          });
        }
      }, 300);
    }
  }, [selectedReqIndex]);

  return (
    <NetworkContext.Provider
      value={value}
      {...props}
    />
  );
};

NetworkProvider.propTypes = {
  autoHighlightChange: PropTypes.bool,
  data: PropTypes.object,
  fetchOptions: PropTypes.object,
  file: PropTypes.string,
  initialState: PropTypes.object,
  onDataError: PropTypes.func,
  onDataLoaded: PropTypes.func,
  onPause: PropTypes.func,
  onReset: PropTypes.func,
  onResume: PropTypes.func,
  scrollRequestPosition: PropTypes.oneOf(['before', 'after', 'near']),
  scrollTimeStamp: PropTypes.number,
};

NetworkProvider.defaultProps = {
  autoHighlightChange: false,
  data: null,
  fetchOptions: { withCredentials: true },
  file: null,
  initialState: defaultState,
  onDataError: null,
  onDataLoaded: null,
  onPause: null,
  onReset: null,
  onResume: null,
  scrollRequestPosition: 'near',
  scrollTimeStamp: null,
};

export default NetworkProvider;
