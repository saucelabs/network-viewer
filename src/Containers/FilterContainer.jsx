import React from 'react';
import { Row, Col } from 'react-flexbox-grid';
import classNames from 'classnames/bind';

import ImportHar from './../Components/Import/ImportHAR';
import Search from './../Components/Filters/Search';
import { useNetwork } from './../state/network/Context';
import { FILTERS } from './../constants';
import Styles from './FilterContainer.styles.scss';
import Button from './../Components/Common/Button';
import { useTheme } from '../state/theme/Context';
import ErrorFilter from '../Components/Filters/ErrorFilter';
import Reset from '../Components/Import/Reset';

const context = classNames.bind(Styles);

const FilterContainer = () => {
  const { state, actions } = useNetwork();
  const { showImportHAR } = useTheme();
  const filter = state.get('filter');
  const filterByError = state.get('errorFilter');

  return (
    <section className={Styles['filters-container']}>
      <Row>
        <Col
          md={5}
          sm={4}
          xs={12}
        >
          <Search
            {...state.get('search')}
            onChange={actions.updateSearch}
          />
        </Col>
        <Col
          md={7}
          sm={8}
          xs={12}
        >
          <div className={Styles['filters-button-group']}>
            {FILTERS.map(({ name, filterBy }) => {
              const selectedFilter = filterBy.value === filter.value;
              const buttonStyle = context('filter-button', {
                'selected-filter': selectedFilter,
              });
              return (
                <Button
                  key={name}
                  category="default"
                  className={buttonStyle}
                  material
                  onClick={() => actions.updateFilter(filterBy)}
                  raised={selectedFilter}
                  size="sm"
                >
                  {name}
                </Button>
              );
            })}
            <ErrorFilter
              isError={filterByError}
              onChange={actions.updateErrorFilter}
            />
            {showImportHAR && (
              <>
                <ImportHar className={Styles['addon-action-button']} />
                <Reset
                  className={Styles['addon-action-button']}
                  onReset={actions.resetState}
                />
              </>
            )}
          </div>
        </Col>
      </Row>
    </section>
  );
};

export default FilterContainer;
