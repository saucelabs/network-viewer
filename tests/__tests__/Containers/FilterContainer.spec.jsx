import React from 'react';
import { mount } from 'enzyme';

import FilterContainer from './../../../src/Containers/FilterContainer';
import NetworkProvider from './../../../src/state/network/NetworkProvider';

describe('FilterContainer', () => {
  it('renders without crashing', () => {
    const element = mount(
      <NetworkProvider>
        <FilterContainer />
      </NetworkProvider>,
    );
    expect(element).toMatchSnapshot();
  });
});
