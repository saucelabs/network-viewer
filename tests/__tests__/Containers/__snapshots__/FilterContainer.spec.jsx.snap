// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterContainer renders without crashing 1`] = `
<NetworkProvider
  autoHighlightChange={false}
  data={null}
  fetchOptions={
    Object {
      "withCredentials": true,
    }
  }
  file={null}
  initialState={
    Immutable.Map {
      "search": Object {
        "name": "URL",
        "value": "",
      },
      "error": null,
      "statusFilter": Object {
        "name": "-",
        "value": null,
      },
      "scrollToIndex": null,
      "data": Immutable.List [],
      "selectedReqIndex": null,
      "reqDetail": null,
      "showReqDetail": false,
      "typeFilter": Object {
        "name": null,
        "value": null,
      },
      "rawData": null,
      "totalNetworkTime": null,
      "actualData": Immutable.List [],
      "dataSummary": Immutable.Map {},
      "loading": false,
      "sort": Object {
        "isAcs": true,
        "key": "startedDateTime",
      },
    }
  }
  onDataError={null}
  onDataLoaded={null}
  onPause={null}
  onReset={null}
  onResume={null}
  scrollRequestPosition="near"
  scrollTimeStamp={null}
>
  <FilterContainer>
    <section>
      <div>
        <StatusFilter>
          <Dropdown
            className={null}
            items={
              Array [
                Object {
                  "name": "-",
                  "value": null,
                },
                Object {
                  "name": "1xx",
                  "value": "1",
                },
                Object {
                  "name": "2xx",
                  "value": "2",
                },
                Object {
                  "name": "3xx",
                  "value": "3",
                },
                Object {
                  "name": "4xx",
                  "value": "4",
                },
                Object {
                  "name": "5xx",
                  "value": "5",
                },
              ]
            }
            onChange={[Function]}
            selected={
              Object {
                "name": "-",
                "value": null,
              }
            }
          >
            <span
              className="dropdown-container"
            >
              <Button
                className="dropdown-toggle"
                disabled={false}
                href={null}
                onClick={[Function]}
                type="button"
                variant="default"
              >
                <button
                  className="btn dropdown-toggle"
                  disabled={false}
                  href={null}
                  onClick={[Function]}
                  type="button"
                >
                  Status: -
                  <IconChevronUp
                    className=""
                  >
                    <svg
                      className=""
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clipRule="evenodd"
                        d="M20.7774 16.7833C21.0742 16.4944 21.0742 16.0261 20.7774 15.7372L12.5374 7.71667C12.2406 7.42778 11.7594 7.42778 11.4626 7.71667L3.22259 15.7372C2.9258 16.0261 2.9258 16.4944 3.22259 16.7833C3.51939 17.0722 4.00059 17.0722 4.29738 16.7833L12 9.2859L19.7026 16.7833C19.9994 17.0722 20.4806 17.0722 20.7774 16.7833Z"
                        fillRule="evenodd"
                      />
                    </svg>
                  </IconChevronUp>
                </button>
              </Button>
            </span>
          </Dropdown>
        </StatusFilter>
        <Search
          name="URL"
          value=""
        >
          <input
            onChange={[Function]}
            placeholder="Search by full URL"
            type="text"
            value=""
          />
        </Search>
        <ResetButton>
          <Button
            className={null}
            disabled={false}
            href={null}
            onClick={[Function]}
            type="button"
            variant="default"
          >
            <button
              className="btn"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              <IconReset
                className=""
              >
                <svg
                  className=""
                  height="24"
                  preserveAspectRatio="xMidYMid"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    clipPath="url(#refresh_clip0)"
                  >
                    <path
                      d="M10.8803 -6.08475e-05C10.9939 -0.000234689 11.0668 0.12043 11.0135 0.220324L7.15162 7.46004C7.09684 7.56273 6.95057 7.56643 6.89026 7.46665L5.2476 4.74904C3.3255 6.36294 2.0772 8.75571 2.00218 11.4547C1.88511 15.6666 4.6625 19.2912 8.53796 20.4307C9.06765 20.5865 9.3703 21.1405 9.21394 21.6681C9.05759 22.1957 8.50144 22.4972 7.97175 22.3414C3.24816 20.9525 -0.139855 16.537 0.00294616 11.3995C0.0976061 7.99401 1.72376 4.99084 4.20545 3.02489L2.52185 0.239546C2.46154 0.13977 2.53326 0.0127309 2.65 0.012552L10.8803 -6.08475e-05Z"
                    />
                    <path
                      d="M14.5072 1.73094C14.6489 1.1992 15.1964 0.882516 15.7302 1.0236C20.4906 2.28174 24 6.60218 24 11.7416C24 15.6586 21.9615 19.0998 18.8846 21.0722L20.5095 23.7603C20.5698 23.8601 20.4981 23.9871 20.3813 23.9873L12.151 23.9999C12.0374 24.0001 11.9645 23.8794 12.0178 23.7796L15.8797 16.5398C15.9345 16.4371 16.0808 16.4335 16.1411 16.5332L17.8526 19.3647C20.35 17.7411 22 14.9325 22 11.7416C22 7.52812 19.123 3.98143 15.2173 2.94919C14.6835 2.8081 14.3656 2.26267 14.5072 1.73094Z"
                    />
                  </g>
                  <defs>
                    <clipPath
                      id="refresh_clip0"
                    >
                      <rect
                        height="24"
                        width="24"
                      />
                    </clipPath>
                  </defs>
                </svg>
              </IconReset>
            </button>
          </Button>
        </ResetButton>
        <ImportHar
          className={null}
          showButton={true}
        >
          <div
            onBlur={[Function]}
            onClick={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
          >
            <input
              accept=".har"
              autoComplete="off"
              multiple={false}
              onChange={[Function]}
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              tabIndex={-1}
              type="file"
            />
            <ImportHarButton>
              <Button
                className={null}
                disabled={false}
                href={null}
                type="button"
                variant="default"
              >
                <button
                  className="btn"
                  disabled={false}
                  href={null}
                  type="button"
                >
                  <IconUpload
                    className=""
                  >
                    <svg
                      className=""
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M9 16.8C9 16.9105 9.08954 17 9.2 17H14.8C14.9105 17 15 16.9105 15 16.8V8H18.8C18.974 8 19.0651 7.78851 18.9475 7.65731L12.1477 0.0663729C12.0684 -0.0221243 11.932 -0.0221243 11.8527 0.0663729L5.05292 7.65731C4.93539 7.78851 5.0264 8 5.20039 8H9V16.8Z"
                      />
                      <path
                        d="M2.5 19.25C2.5 18.5596 1.94036 18 1.25 18C0.559644 18 0 18.5596 0 19.25V22.75C0 23.4404 0.559644 24 1.25 24H22.75C23.4404 24 24 23.4404 24 22.75V19.25C24 18.5596 23.4404 18 22.75 18C22.0596 18 21.5 18.5596 21.5 19.25V21.5H2.5V19.25Z"
                      />
                    </svg>
                  </IconUpload>
                </button>
              </Button>
            </ImportHarButton>
          </div>
        </ImportHar>
      </div>
      <div>
        <TypeFilter>
          <Button
            className="active"
            disabled={false}
            href={null}
            key="All"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn active text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              All
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="XHR"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              XHR
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="JS"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              JS
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="CSS"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              CSS
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="Img"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              Img
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="Media"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              Media
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="Font"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              Font
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="Doc"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              Doc
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="WS"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              WS
            </button>
          </Button>
          <Button
            className=""
            disabled={false}
            href={null}
            key="Manifest"
            onClick={[Function]}
            type="button"
            variant="text"
          >
            <button
              className="btn text-variant"
              disabled={false}
              href={null}
              onClick={[Function]}
              type="button"
            >
              Manifest
            </button>
          </Button>
        </TypeFilter>
      </div>
    </section>
  </FilterContainer>
</NetworkProvider>
`;
